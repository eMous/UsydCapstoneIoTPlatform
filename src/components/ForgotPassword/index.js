import React, { useState } from "react";
import Grid from "@material-ui/core/Grid";
import Paper from "@material-ui/core/Paper";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";
import MuiAlert from "@material-ui/lab/Alert";
import Snackbar from "@material-ui/core/Snackbar";
import useStyles from "./styles";
import resetUserPassword from "./services/sendPasswordReset";

const ForgotPassword = () => {
  const classes = useStyles();

  const [emailAddr, setEmailAddr] = useState("");
  const [isSnackbarOpen, setIsSnackbarOpen] = useState(false);
  const [errorMsg, setErrorMsg] = useState("");
  const [isBtnDisabled, setIsBtnDisabled] = useState(false);

  const closeSnackbar = (event, reason) => {
    if (reason === "clickaway") {
      return;
    }

    setIsSnackbarOpen(false);
  };

  const btnDebounce = async () => {
    try {
      // Attempt to make a password reset request
      await resetUserPassword(emailAddr);

      // If successful, we generate a success message to the user
      setIsSnackbarOpen(true);
      setErrorMsg("");

      // We disable the button
      setIsBtnDisabled(true);

      // After a fixed amount of time, we enable the button again
      setTimeout(() => {
        setIsBtnDisabled(false);
      }, 5000);
    } catch (error) {
      // Otherwise if there's an error, display the error message to the user
      setErrorMsg(error.toString());
    }
  };

  return (
    <Paper className={classes.paper} variant="outlined">
      <form className={classes.textFieldRoot}>
        <Grid container direction="column" spacing={4}>
          <Grid item>
            <Typography variant="h5">Forgot Password</Typography>
          </Grid>
          <Grid item>
            <TextField
              required
              variant="outlined"
              id="email-addr"
              label="Email address"
              type="email"
              value={emailAddr}
              onChange={(e) => {
                setEmailAddr(e.target.value);
              }}
              error={!!errorMsg}
              helperText={errorMsg}
              fullWidth
            />
          </Grid>
          <Grid item>
            <Button
              variant="contained"
              color="secondary"
              disabled={isBtnDisabled}
              onClick={btnDebounce}
            >
              Reset my password
            </Button>
          </Grid>
        </Grid>
        <Snackbar
          open={isSnackbarOpen}
          autoHideDuration={5000}
          onClose={closeSnackbar}
        >
          <MuiAlert
            elevation={6}
            variant="filled"
            onClose={closeSnackbar}
            severity="success"
          >
            Your password reset link has been sent to your email address!
          </MuiAlert>
        </Snackbar>
      </form>
    </Paper>
  );
};

export default ForgotPassword;
