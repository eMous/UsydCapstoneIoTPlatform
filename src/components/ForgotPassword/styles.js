import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles((theme) => ({
  paper: {
    padding: theme.spacing(4),
    width: "100%",
    maxWidth: "25rem",
    margin: "auto",
  },

  textFieldRoot: {
    "& .MuiTextField-root": {
      marginTop: theme.spacing(2),
      marginBottom: theme.spacing(2),
    },
  },
}));

export default useStyles;
