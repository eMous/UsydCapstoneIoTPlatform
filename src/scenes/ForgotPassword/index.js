import React from "react";
import ForgotPassword from "components/ForgotPassword";
import useStyles from "./styles";

const ForgotPasswordScene = () => {
  const classes = useStyles();

  return (
    <div className={classes.root}>
      <ForgotPassword />
    </div>
  );
};

export default ForgotPasswordScene;
