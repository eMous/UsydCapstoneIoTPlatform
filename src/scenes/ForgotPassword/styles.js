import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles((theme) => ({
  root: {
    backgroundColor: theme.palette.background.default,
    width: "100%",
    height: "100%",
    display: "flex",
    position: "fixed",
    justifyContent: "center",
    alignItems: "center",
    "& > * + *": {
      marginTop: theme.spacing(2),
    },
  },
}));

export default useStyles;
