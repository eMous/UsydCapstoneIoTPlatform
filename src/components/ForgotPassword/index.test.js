import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, screen, render, waitFor } from "@testing-library/react";
import ForgotPassword from "./index";

jest.mock("./services/sendPasswordReset");
import resetUserPassword from "./services/sendPasswordReset";

describe("ForgotPassword component", () => {
  test("Snapshot of component is rendered", () => {
    const component = renderer.create(<ForgotPassword />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("Reset password fails", () => {
    afterAll(() => {
      jest.clearAllMocks();
    });

    test("Error helper text is displayed when password reset has an error", async () => {
      resetUserPassword.mockImplementation(() => {
        throw new Error(
          "Forgot Password Error: The email address provided is invalid."
        );
      });

      render(<ForgotPassword />);

      // Click on the Reset Password button
      await waitFor(() => {
        fireEvent.click(screen.getByText(/Reset my password/i));
      });

      // We expect an error text to appear
      expect(await screen.getByText(/forgot password error/i)).toBeTruthy();
    });
  });

  describe("Reset password succeeds", () => {
    afterAll(() => {
      jest.clearAllMocks();
    });

    test("Successful Snackbar appears after email is successfully sent out", async () => {
      resetUserPassword.mockImplementation(() => jest.fn());
      render(<ForgotPassword />);

      const resetPwdBtn = screen.getByText(/Reset my password/i);

      // Click the reset password button
      fireEvent.click(resetPwdBtn);

      // Wait for the Snackbar to appear
      await waitFor(() => {
        expect(screen.getByText(/Your password reset link/i)).toBeTruthy();
      });
    });

    test("Reset password button is disabled after successfully completing operation", async () => {
      resetUserPassword.mockImplementation(() => jest.fn());
      render(<ForgotPassword />);

      const resetPwdBtn = screen.getByText(/Reset my password/i);

      // Click the reset password button
      fireEvent.click(resetPwdBtn);

      // Wait for the button to be disabled
      await waitFor(() => {
        expect(
          screen.getByText(/Reset my password/i).closest("button")
        ).toBeDisabled();
      });
    });

    test("Reset password button is re-enabled after being disabled for some time", async () => {
      resetUserPassword.mockImplementation(() => jest.fn());
      render(<ForgotPassword />);

      const resetPwdBtn = screen.getByText(/Reset my password/i);

      // Click the reset password button
      fireEvent.click(resetPwdBtn);

      // Wait for the button to be re-enabled
      await waitFor(() => {
        expect(
          screen.getByText(/Reset my password/i).closest("button")
        ).not.toBeDisabled();
      });
    });

    test("Successful Snackbar disappears after some time", async () => {
      resetUserPassword.mockImplementation(() => jest.fn());

      render(<ForgotPassword />);

      const resetPwdBtn = screen.getByText(/Reset my password/i);

      // Click the reset password button
      fireEvent.click(resetPwdBtn);

      await waitFor(() => {
        expect(screen.queryByText(/Your password reset link/i)).toBeNull;
      });
    });
  });
});
