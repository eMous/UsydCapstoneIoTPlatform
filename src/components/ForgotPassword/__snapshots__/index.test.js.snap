// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword component Snapshot of component is rendered 1`] = `
<div
  className="MuiPaper-root makeStyles-paper-1 MuiPaper-outlined MuiPaper-rounded"
>
  <form
    className="makeStyles-textFieldRoot-2"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-direction-xs-column"
    >
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <h5
          className="MuiTypography-root MuiTypography-h5"
        >
          Forgot Password
        </h5>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
            data-shrink={false}
            htmlFor="email-addr"
            id="email-addr-label"
          >
            Email address
            <span
              aria-hidden={true}
              className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              id="email-addr"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="email"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-5"
              >
                <span>
                  Email address
                   *
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Reset my password
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;
